File: services/elixir-phoenix/lib/stt_gateway/inference.ex
defmodule SttGateway.Inference do
  @moduledoc false

  alias SttGateway.Protocol.ErrorMessage

  require Logger

  @type partial_context :: %{
          oldest_seq: non_neg_integer(),
          newest_seq: non_neg_integer(),
          frames: non_neg_integer(),
          flush_lateness_ms: float(),
          oldest_received_at_us: integer(),
          newest_received_at_us: integer()
        }

  @type inference_result :: %{
          rms: float(),
          zero_crossings: non_neg_integer(),
          checksum: non_neg_integer(),
          samples: non_neg_integer(),
          transcript: String.t(),
          audio_bytes: non_neg_integer()
        }

  @type error_payload :: %{
          stage: ErrorMessage.stage(),
          kind: ErrorMessage.kind(),
          message: String.t(),
          inference_status: non_neg_integer() | nil,
          inference_elapsed_ms: float() | nil,
          retryable: boolean()
        }

  @type result :: {:ok, inference_result()} | {:error, error_payload()}

  @spec request_async(iodata(), non_neg_integer(), partial_context(), pid(), String.t(), atom()) ::
          :ok
  def request_async(body, cpu_passes, context, caller, inference_url, finch_name) do
    Task.start(fn -> run(body, cpu_passes, context, caller, inference_url, finch_name) end)
    :ok
  end

  @spec run(iodata(), non_neg_integer(), partial_context(), pid(), String.t(), atom()) :: :ok
  defp run(body, cpu_passes, context, caller, inference_url, finch_name) do
    request =
      Finch.build(
        :post,
        "#{inference_url}/infer",
        [{"x-cpu-passes", Integer.to_string(cpu_passes)}],
        body
      )

    started_at = System.monotonic_time(:microsecond)
    result = call_inference(request, finch_name, started_at)
    send(caller, {:compute_finished, result, context})
    :ok
  end

  @spec call_inference(Finch.Request.t(), atom(), integer()) :: result()
  defp call_inference(request, finch_name, started_at) do
    case Finch.request(request, finch_name, receive_timeout: 2_000, pool_timeout: 1_000) do
      {:ok, %Finch.Response{status: 200, body: body}} ->
        decode_body(body, started_at)

      {:ok, %Finch.Response{status: status}} ->
        Logger.warning("inference returned status #{status}")
        {:error, status_error(status, started_at)}

      {:error, error} ->
        Logger.warning("inference call failed: #{inspect(error)}")
        {:error, transport_error(error, started_at)}
    end
  end

  @spec decode_body(binary(), integer()) :: result()
  defp decode_body(body, started_at) do
    case JSON.decode(body) do
      {:ok, decoded} ->
        {:ok, build_result(decoded)}

      {:error, decode_error} ->
        Logger.warning("inference response decode failed: #{inspect(decode_error)}")

        {:error,
         make_error_payload(
           :inference_response_parse,
           :parse_error,
           inspect(decode_error),
           200,
           started_at,
           false
         )}
    end
  end

  @spec status_error(non_neg_integer(), integer()) :: error_payload()
  defp status_error(status, started_at) do
    {kind, retryable} =
      cond do
        status == 429 -> {:http_429, true}
        status in 500..599 -> {:http_5xx, true}
        true -> {:http_5xx, false}
      end

    make_error_payload(
      :inference_request,
      kind,
      "inference returned status #{status}",
      status,
      started_at,
      retryable
    )
  end

  @spec transport_error(term(), integer()) :: error_payload()
  defp transport_error(error, started_at) do
    {kind, message} =
      case error do
        %Mint.TransportError{reason: :timeout} -> {:timeout, "inference request timed out"}
        _ -> {:connection_reset, inspect(error)}
      end

    make_error_payload(:inference_request, kind, message, nil, started_at, true)
  end

  @spec make_error_payload(
          ErrorMessage.stage(),
          ErrorMessage.kind(),
          String.t(),
          non_neg_integer() | nil,
          integer(),
          boolean()
        ) :: error_payload()
  defp make_error_payload(stage, kind, message, inference_status, started_at, retryable) do
    %{
      stage: stage,
      kind: kind,
      message: message,
      inference_status: inference_status,
      inference_elapsed_ms: elapsed_ms_since(started_at),
      retryable: retryable
    }
  end

  @spec build_result(map()) :: inference_result()
  defp build_result(decoded) do
    %{
      rms: decoded["rms"],
      zero_crossings: decoded["zero_crossings"],
      checksum: decoded["checksum"],
      samples: decoded["samples"],
      transcript: decoded["transcript"],
      audio_bytes: decoded["audio_bytes"]
    }
  end

  @spec elapsed_ms_since(integer()) :: float()
  defp elapsed_ms_since(started_at_us) do
    (System.monotonic_time(:microsecond) - started_at_us) / 1000.0
  end
end

File: services/rust-axum/src/inference.rs
use std::time::Instant;

use bytes::{Bytes, BytesMut};

use crate::config::{CPU_PASSES_HEADER, FRAME_BYTES};
use crate::protocol::{
    ERROR_TYPE, ErrorKind, ErrorMessage, ErrorStage, Frame, InferResponse, PartialContext,
    PartialMessage,
};
use crate::state::AppContext;

pub(crate) struct InferenceContext<F: Fn() -> usize> {
    pub(crate) flush_lateness_ms: f64,
    /// Sampler the helper calls *only on the error path* to read
    /// `gateway_buffer_frames`. Avoids locking the buffer mutex on the
    /// happy path where the value is unused.
    pub(crate) gateway_buffer_frames_at_error: F,
}

pub(crate) async fn request_partial<F: Fn() -> usize>(
    context: &AppContext,
    batch: Vec<Frame>,
    inference_context: InferenceContext<F>,
) -> Result<PartialMessage, ErrorMessage> {
    let oldest = batch.first().expect("non-empty batch");
    let newest = batch.last().expect("non-empty batch");
    let oldest_seq = oldest.seq;
    let newest_seq = newest.seq;
    let oldest_received_at = oldest.received_at;
    let newest_received_at = newest.received_at;
    let frames = batch.len();
    let body = concat_pcm(&batch);
    let body_len = body.len() as u64;
    let partial_context = PartialContext {
        oldest_frame_seq: oldest_seq,
        newest_frame_seq: newest_seq,
        frames,
        flush_lateness_ms: inference_context.flush_lateness_ms,
    };

    let started_at = Instant::now();
    let make_error = |stage: ErrorStage,
                      kind: ErrorKind,
                      message: String,
                      inference_status: Option<u16>,
                      retryable: bool|
     -> ErrorMessage {
        let now = Instant::now();
        ErrorMessage {
            r#type: ERROR_TYPE,
            stage,
            kind,
            message,
            oldest_frame_seq: oldest_seq,
            newest_frame_seq: newest_seq,
            frames,
            audio_bytes: body_len,
            oldest_age_ms: duration_ms(now, oldest_received_at),
            newest_age_ms: duration_ms(now, newest_received_at),
            flush_lateness_ms: inference_context.flush_lateness_ms,
            inference_elapsed_ms: Some(duration_ms(now, started_at)),
            // Always 1: the one-inflight-per-connection invariant from CLAUDE.md
            // means we *are* the inflight batch. The field is on-wire so clients
            // can reason about gateway pressure under future relaxed designs.
            inflight_gateway_batches: 1,
            gateway_buffer_frames: (inference_context.gateway_buffer_frames_at_error)(),
            inference_status,
            retryable,
        }
    };

    let response = match context
        .inference_client()
        .post(context.inference_url.as_ref())
        .header(CPU_PASSES_HEADER, context.cpu_passes_header.clone())
        .body(body)
        .send()
        .await
    {
        Ok(response) => response,
        Err(error) => {
            let (kind, retryable) = classify_request_error(&error);
            return Err(make_error(
                ErrorStage::InferenceRequest,
                kind,
                error.to_string(),
                None,
                retryable,
            ));
        }
    };

    let status = response.status();
    if !status.is_success() {
        let (kind, retryable) = classify_status(status.as_u16());
        return Err(make_error(
            ErrorStage::InferenceRequest,
            kind,
            format!("inference returned status {status}"),
            Some(status.as_u16()),
            retryable,
        ));
    }

    let infer: InferResponse = match response.json().await {
        Ok(infer) => infer,
        Err(error) => {
            return Err(make_error(
                ErrorStage::InferenceResponseParse,
                ErrorKind::ParseError,
                error.to_string(),
                Some(status.as_u16()),
                false,
            ));
        }
    };

    Ok(PartialMessage::new(
        infer,
        partial_context,
        context.cpu_passes,
        context.model_delay,
    ))
}

fn concat_pcm(batch: &[Frame]) -> Bytes {
    let mut body = BytesMut::with_capacity(batch.len() * FRAME_BYTES);
    for frame in batch {
        body.extend_from_slice(&frame.payload);
    }
    body.freeze()
}

fn duration_ms(now: Instant, earlier: Instant) -> f64 {
    now.saturating_duration_since(earlier).as_secs_f64() * 1000.0
}

fn classify_request_error(error: &reqwest::Error) -> (ErrorKind, bool) {
    if error.is_timeout() {
        return (ErrorKind::Timeout, true);
    }
    if error.is_connect() {
        return (ErrorKind::ConnectionReset, true);
    }
    if error.is_decode() {
        return (ErrorKind::ParseError, false);
    }
    (ErrorKind::ConnectionReset, true)
}

fn classify_status(status: u16) -> (ErrorKind, bool) {
    if status == 429 {
        return (ErrorKind::Http429, true);
    }
    if (500..600).contains(&status) {
        return (ErrorKind::Http5xx, true);
    }
    (ErrorKind::Http5xx, false)
}
